File: python/src/anysearch/providers/you.py
"""You.com (YDC) Search API — web results with rich snippets."""

from __future__ import annotations

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability


class YouProvider(BaseProvider):
    name = "you"
    aliases = ("youcom", "ydc")
    env_keys = ("YDC_API_KEY", "YOU_API_KEY")
    default_base_url = "https://ydc-index.io"
    capabilities = frozenset(
        {
            Capability.DOMAINS,
            Capability.COUNTRY,
            Capability.HIGHLIGHTS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        body = {"query": req.query, "count": req.max_results}
        if req.include_domains:
            body["include_domains"] = req.include_domains
        if req.exclude_domains:
            body["exclude_domains"] = req.exclude_domains
        if req.country:
            body["country"] = req.country.lower()
        if req.safe_search:
            body["safesearch"] = req.safe_search
        body.update(req.extra)
        return PreparedRequest(
            method="POST",
            url=f"{self.base_url}/v1/search",
            headers={"X-API-Key": self.api_key, "Content-Type": "application/json"},
            json=body,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        items = data.get("results") or data.get("hits") or []
        results = []
        for item in items:
            snippets = item.get("snippets") or []
            authors = item.get("authors") or []
            description = item.get("description")
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=description or (snippets[0] if snippets else None),
                    text="\n\n".join(snippets) if snippets else None,
                    highlights=list(snippets),
                    published_date=item.get("page_age"),
                    author=authors[0] if authors else None,
                    raw=item,
                )
            )
        return self._response(req, results, data, elapsed_ms=elapsed_ms)

File: js/src/providers/you.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

export const you: ProviderSpec = {
  name: "you",
  aliases: ["youcom", "ydc"],
  envKeys: ["YDC_API_KEY", "YOU_API_KEY"],
  defaultBaseUrl: "https://ydc-index.io",
  capabilities: new Set(["domains", "country", "highlights"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const body: Record<string, unknown> = { query: req.query, count: req.maxResults };
    if (req.includeDomains.length) body.include_domains = req.includeDomains;
    if (req.excludeDomains.length) body.exclude_domains = req.excludeDomains;
    if (req.country) body.country = req.country.toLowerCase();
    if (req.safeSearch) body.safesearch = req.safeSearch;
    Object.assign(body, req.extra);
    return {
      method: "POST",
      url: `${ctx.baseUrl}/v1/search`,
      headers: { "X-API-Key": ctx.apiKey ?? "", "Content-Type": "application/json" },
      json: body,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const items = data.results ?? data.hits ?? [];
    const results = items.map((item: any) => {
      const snippets: string[] = item.snippets ?? [];
      const authors: string[] = item.authors ?? [];
      return result({
        title: item.title,
        url: item.url,
        snippet: item.description ?? snippets[0],
        text: snippets.length ? snippets.join("\n\n") : undefined,
        highlights: snippets,
        publishedDate: item.page_age,
        author: authors[0],
        raw: item,
      });
    });
    return finalize(you, req, results, data, { elapsedMs });
  },
};
